# Machine Learning Model with The Data of Clinics in Lower Saxony Germany



![Machine Learning Workflow](Workflow_ML.png)

#### Erforderliche Importe,die benötigt werden.

In [708]:
import pandas as pd
import numpy as np
import re
#import csv

import textblob
from textblob import TextBlob
from textblob_de import TextBlobDE
from textblob_de import PatternParser
from textblob.sentiments import NaiveBayesAnalyzer

import nltk
from nltk.corpus import stopwords
from nltk.stem.snowball import SnowballStemmer
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords

from wordcloud import WordCloud, STOPWORDS

import matplotlib.pyplot as plt
import matplotlib as mpl
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
%matplotlib inline

from PIL import Image
import pickle

import cufflinks as cf
cf.go_offline()
cf.set_config_file(offline=False, world_readable=True)
import bar_chart_race as bcr
from pathlib import Path

In [709]:
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.svm import LinearSVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.neural_network import MLPClassifier
from xgboost import XGBClassifier
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV

#from bokeh.io import output_notebook
output_notebook()

pd.options.display.max_columns = 100
pd.options.display.max_rows = 300
pd.options.display.max_colwidth = 100
np.set_printoptions(threshold=2000)
#import holoviews as hv
#import hvplot
#from hvplot import pandas
#hv.extension("bokeh")

Loading BokehJS ...

##### Konvertieren csv-Daten in Pandas DataFrame

In [710]:
data = pd.read_csv(r'C:\Users\name\Desktop\WB\Abschlussproject\klinikbewertung.de\alldata\alldata_klinikandgoogle.csv', encoding='utf-8')

#### Wir erstellen Dataframe

In [711]:
df = pd.DataFrame(data, columns= ['Name der Klinik', 'Textuelle Bewertung','Sternebewertung','Rating'])
df

Name der Klinik  \
0        Augenklinik Dr.Hoffmann   
1        Augenklinik Dr.Hoffmann   
2        Augenklinik Dr.Hoffmann   
3        Augenklinik Dr.Hoffmann   
4        Augenklinik Dr.Hoffmann   
...                          ...   
5231  Kreiskrankenhaus Osterholz   
5232  Kreiskrankenhaus Osterholz   
5233  Kreiskrankenhaus Osterholz   
5234  Kreiskrankenhaus Osterholz   
5235  Kreiskrankenhaus Osterholz   

                                                                                      Textuelle Bewertung  \
0     Meine Laser OP grauer Star hat super geklappt. Alles lief wie am Schnürchen ab. Danke dafür.Dank...   
1     Bin sehr zufrieden gewesen mit der Klinik und dem Personal. Die erste OP verlief sehr gut +die  ...   
2                       Komptentes Ärzteteam.Strickte Einhaltung der Corona Schutzregeln.Rundum zufrieden   
3     Ich komme seit einen halben Jahr mit meiner behinderten Frau hier zum Augen-Spritzen auf Grund e...   
4     Ich war Anfang November 2019 zur Augen-OP (Grauer Star)in der Klinik und zunächst wegen negative...   
...                                                                                                   ...   
5231  Ambulante Versorgung bei meinem Ehemann, waren freundlich zugewandt und keine langen Wartezeiten...   
5232  Sehr gute Ärzte war am 20.04.2018 zur ambulanten OP im Krankenhaus Osterholz Scharmbeck von der ...   
5233                                                 Schönes kleines Krankenhaus mit sehr nettem Personal   
5234                                                               Gute Pflege in einem guten Krankenhaus   
5235                                               Gynäkologie Arzt sehr unfreundlich ,schlechte Beratung   

      Sternebewertung  Rating  
0                 6.0     NaN  
1                 6.0     NaN  
2                 6.0     NaN  
3                 6.0     NaN  
4                 6.0     NaN  
...               ...     ...  
5231              3.0     NaN  
5232              5.0     NaN  
5233              5.0     NaN  
5234              5.0     NaN  
5235              1.0     NaN  

[5236 rows x 4 columns]

In [712]:
#for q in range(result.shape[0]):
# create a list of our conditions
conditions = [
    (df['Sternebewertung'] >= 3 ),
    (df['Sternebewertung'] < 3 )
    ]
# create a list of the values we want to assign for each condition
values = [1, 0]

# create a new column and use np.select to assign values to it using our lists as arguments
df['Rating'] = np.select(conditions, values)

# display updated DataFrame
df.head(50)

Name der Klinik  \
0                Augenklinik Dr.Hoffmann   
1                Augenklinik Dr.Hoffmann   
2                Augenklinik Dr.Hoffmann   
3                Augenklinik Dr.Hoffmann   
4                Augenklinik Dr.Hoffmann   
5                Augenklinik Dr.Hoffmann   
6                Augenklinik Dr.Hoffmann   
7                Augenklinik Dr.Hoffmann   
8                Augenklinik Dr.Hoffmann   
9                Augenklinik Dr.Hoffmann   
10               Augenklinik Dr.Hoffmann   
11               Augenklinik Dr.Hoffmann   
12               Augenklinik Dr.Hoffmann   
13               Augenklinik Dr.Hoffmann   
14               Augenklinik Dr.Hoffmann   
15               Augenklinik Dr.Hoffmann   
16               Augenklinik Dr.Hoffmann   
17               Augenklinik Dr.Hoffmann   
18               Augenklinik Dr.Hoffmann   
19               Augenklinik Dr.Hoffmann   
20               Augenklinik Dr.Hoffmann   
21               Augenklinik Dr.Hoffmann   
22               Augenklinik Dr.Hoffmann   
23  Krankenhaus Marienstift Braunschweig   
24  Krankenhaus Marienstift Braunschweig   
25  Krankenhaus Marienstift Braunschweig   
26  Krankenhaus Marienstift Braunschweig   
27  Krankenhaus Marienstift Braunschweig   
28  Krankenhaus Marienstift Braunschweig   
29  Krankenhaus Marienstift Braunschweig   
30  Krankenhaus Marienstift Braunschweig   
31  Krankenhaus Marienstift Braunschweig   
32  Krankenhaus Marienstift Braunschweig   
33  Krankenhaus Marienstift Braunschweig   
34  Krankenhaus Marienstift Braunschweig   
35  Krankenhaus Marienstift Braunschweig   
36  Krankenhaus Marienstift Braunschweig   
37  Krankenhaus Marienstift Braunschweig   
38  Krankenhaus Marienstift Braunschweig   
39  Krankenhaus Marienstift Braunschweig   
40  Krankenhaus Marienstift Braunschweig   
41  Krankenhaus Marienstift Braunschweig   
42  Krankenhaus Marienstift Braunschweig   
43  Krankenhaus Marienstift Braunschweig   
44  Krankenhaus Marienstift Braunschweig   
45  Krankenhaus Marienstift Braunschweig   
46  Krankenhaus Marienstift Braunschweig   
47  Krankenhaus Marienstift Braunschweig   
48  Krankenhaus Marienstift Braunschweig   
49  Krankenhaus Marienstift Braunschweig   

                                                                                    Textuelle Bewertung  \
0   Meine Laser OP grauer Star hat super geklappt. Alles lief wie am Schnürchen ab. Danke dafür.Dank...   
1   Bin sehr zufrieden gewesen mit der Klinik und dem Personal. Die erste OP verlief sehr gut +die  ...   
2                     Komptentes Ärzteteam.Strickte Einhaltung der Corona Schutzregeln.Rundum zufrieden   
3   Ich komme seit einen halben Jahr mit meiner behinderten Frau hier zum Augen-Spritzen auf Grund e...   
4   Ich war Anfang November 2019 zur Augen-OP (Grauer Star)in der Klinik und zunächst wegen negative...   
5   Trotz extrem hoher Patientenfrequenz erfolgte die Behandlung/Untersuchung in angemessener Zeit. ...   
6   Diese Klinik ist NUR zu empfehlen!!! Das gesamte Team: Ärzte, Helferinnen, Pflege- und Hauswirts...   
7   Sehr nettes und kompetentens Personal!! Kurze Wartezeiten trotz hoher Frequentierung. Gute Aufkl...   
8                                   Freundliches Personal, keine lange Wartezeit, kompetente Behandlung   
9   Dass es bei einer offenen Sprechstunde einer Augenklinik zu langen Wartezeiten kommt, ist verstä...   
10  Die Schwestern in der Anmeldung habe ich zum wiederholten Male als sehr unfreundlich (genervt) u...   
11  Ich habe die Klinik aufgesucht, um schnell eine Diagnose zu bekommen. Andere Augenärzte wollten ...   
12  In der Anmeldung ist Chaos...Nachdem ich 1 Std. im Wartezimmer saß, meldete ich mich erneut bei ...   
13  Hallo, telefonische Beratung war inkompetent. Ich habe einen Termin vereinbart und habe gefragt,...   
14  Hut ab! Am 19. Februar 2017 kam ich als Notfall in die Klinik. Bereits am Empfang habe ich freun...   
15  Ich bin an Diabetes erkrankt und habe leider auch schon Folgesch

In [713]:
df.dtypes

Name der Klinik         object
Textuelle Bewertung     object
Sternebewertung        float64
Rating                   int32
dtype: object

In [714]:
df.info

<bound method DataFrame.info of                  Name der Klinik  \
0        Augenklinik Dr.Hoffmann   
1        Augenklinik Dr.Hoffmann   
2        Augenklinik Dr.Hoffmann   
3        Augenklinik Dr.Hoffmann   
4        Augenklinik Dr.Hoffmann   
...                          ...   
5231  Kreiskrankenhaus Osterholz   
5232  Kreiskrankenhaus Osterholz   
5233  Kreiskrankenhaus Osterholz   
5234  Kreiskrankenhaus Osterholz   
5235  Kreiskrankenhaus Osterholz   

                                                                                      Textuelle Bewertung  \
0     Meine Laser OP grauer Star hat super geklappt. Alles lief wie am Schnürchen ab. Danke dafür.Dank...   
1     Bin sehr zufrieden gewesen mit der Klinik und dem Personal. Die erste OP verlief sehr gut +die  ...   
2                       Komptentes Ärzteteam.Strickte Einhaltung der Corona Schutzregeln.Rundum zufrieden   
3     Ich komme seit einen halben Jahr mit meiner behinderten Frau hier zum Augen-Spritzen auf Grun

## Preproccessing and Data Preparation

Obwohl wir die Daten als UTF-8 kodiert haben, sollen wir einige Falsche Wörter ändern.

In [715]:
nltk.download('punkt')
nltk.download('stopwords')

stemmer = SnowballStemmer("german")
stop_words = set(stopwords.words("german"))

def clean_text(text, for_embedding=False):
    """
        - remove any html tags (< /br> often found)
        - Keep only ASCII + European Chars and whitespace, no digits
        - remove single letter chars
        - convert all whitespaces (tabs etc.) to single wspace
        if not for embedding (but e.g. tdf-idf):
        - all lowercase
        - remove stopwords, punctuation and stemm
        - some encoding errors 
    """
    RE_WSPACE = re.compile(r"\s+", re.IGNORECASE)
    RE_TAGS = re.compile(r"<[^>]+>")
    RE_ASCII = re.compile(r"[^A-Za-zÀ-ž ]", re.IGNORECASE)
    RE_SINGLECHAR = re.compile(r"\b[A-Za-zÀ-ž]\b", re.IGNORECASE)
    if for_embedding:
        # Keep punctuation
        RE_ASCII = re.compile(r"[^A-Za-zÀ-ž,.!? ]", re.IGNORECASE)
        RE_SINGLECHAR = re.compile(r"\b[A-Za-zÀ-ž,.!?]\b", re.IGNORECASE)

    text = re.sub(RE_TAGS, " ", text)
    text = re.sub(RE_ASCII, " ", text)
    text = re.sub(RE_SINGLECHAR, " ", text)
    text = re.sub(RE_WSPACE, " ", text)
    #text= text.str.replace('\n','')
    #text= text.str.replace('ö','ö')
    #text= text.str.replace('Ã\x84','Ä')
    #text= text.str.replace('ä','ä')
    #text= text.str.replace('ü','ü')
    #text= text.str.replace('Ã\x9f','ß')
    #text= text.str.replace('Ã\x9c','Ü')

    word_tokens = word_tokenize(text)
    words_tokens_lower = [word.lower() for word in word_tokens]

    if for_embedding:
        # no stemming, lowering and punctuation / stop words removal
        words_filtered = word_tokens
    else:
        words_filtered = [
            stemmer.stem(word) for word in words_tokens_lower if word not in stop_words
        ]

    text_clean = " ".join(words_filtered)
    return text_clean

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\name\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\name\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


# Daten Modell

In [716]:
df

Name der Klinik  \
0        Augenklinik Dr.Hoffmann   
1        Augenklinik Dr.Hoffmann   
2        Augenklinik Dr.Hoffmann   
3        Augenklinik Dr.Hoffmann   
4        Augenklinik Dr.Hoffmann   
...                          ...   
5231  Kreiskrankenhaus Osterholz   
5232  Kreiskrankenhaus Osterholz   
5233  Kreiskrankenhaus Osterholz   
5234  Kreiskrankenhaus Osterholz   
5235  Kreiskrankenhaus Osterholz   

                                                                                      Textuelle Bewertung  \
0     Meine Laser OP grauer Star hat super geklappt. Alles lief wie am Schnürchen ab. Danke dafür.Dank...   
1     Bin sehr zufrieden gewesen mit der Klinik und dem Personal. Die erste OP verlief sehr gut +die  ...   
2                       Komptentes Ärzteteam.Strickte Einhaltung der Corona Schutzregeln.Rundum zufrieden   
3     Ich komme seit einen halben Jahr mit meiner behinderten Frau hier zum Augen-Spritzen auf Grund e...   
4     Ich war Anfang November 2019 zur Augen-OP (Grauer Star)in der Klinik und zunächst wegen negative...   
...                                                                                                   ...   
5231  Ambulante Versorgung bei meinem Ehemann, waren freundlich zugewandt und keine langen Wartezeiten...   
5232  Sehr gute Ärzte war am 20.04.2018 zur ambulanten OP im Krankenhaus Osterholz Scharmbeck von der ...   
5233                                                 Schönes kleines Krankenhaus mit sehr nettem Personal   
5234                                                               Gute Pflege in einem guten Krankenhaus   
5235                                               Gynäkologie Arzt sehr unfreundlich ,schlechte Beratung   

      Sternebewertung  Rating  
0                 6.0       1  
1                 6.0       1  
2                 6.0       1  
3                 6.0       1  
4                 6.0       1  
...               ...     ...  
5231              3.0       1  
5232              5.0       1  
5233              5.0       1  
5234              5.0       1  
5235              1.0       0  

[5236 rows x 4 columns]

In [717]:
"""Compute unique word vector with frequencies
exclude very uncommon (<10 obsv.) and common (>=30%) words
use pairs of two words (ngram)
"""
vectorizer = TfidfVectorizer(
    analyzer="word", max_df=0.3, min_df=10, ngram_range=(1, 2), norm="l2"
)
vectorizer.fit(df["Textuelle Bewertung"])

TfidfVectorizer(max_df=0.3, min_df=10, ngram_range=(1, 2))

An important parameter that needs explanation is the ngram_range. An ngram of one means that you look at each word separately. An ngram of two (or bigram) means that you take the preceding and following word into account as well. Thus, some context is added. This is helpful because then a model can learn that "good" and "not good" are different. In our case, in addition to using each word by itself we also add bigrams to make use of context. Let's see some of the created ngrams and their indices:

In [634]:
# Vector representation of vocabulary
word_vector = pd.Series(vectorizer.vocabulary_).sample(3, random_state=1)
print(f"Unique word (ngram) vector extract:\n\n {word_vector}")

Unique word (ngram) vector extract:

 müssen und        5606
stunden später    7258
über die          9091
dtype: int64


This creates a numeric representation of the ngrams in our corpus. We see that the vectorizer also uses bigrams in addition to single words. The word "müssen und" is represented by the number 5606, while the number 7258 stands for the bigram "stunden später" and so on.
This is only the first part of our text to numeric process. Before we can move on to transform each sentence to a vector of TF-IDF values, we need to prepare the data for the modeling part first.

### Modeling

To test the classification performance of our model, we will perform a cross validation. For that, we split our data into a training and a testing set. The former is used to train the model and the latter to evaluate its predictions:

In [635]:
# Sample data - 25% of data to test set
train, test = train_test_split(result, random_state=1, test_size=0.25, shuffle=True)

X_train = train["Textuelle Bewertung"]
Y_train = train["Rating"]
X_test = test["Textuelle Bewertung"]
Y_test = test["Rating"]
print(X_train.shape)
print(X_test.shape)

(3927,)
(1309,)


The training set consists of more than 3927 rows and the testing will be performed on more than 1309 observations. Now, that we have split our data we can transform the text data into its `TF-IDF` representation:

In [636]:
# transform each sentence to numeric vector with tf-idf value as elements
X_train_vec = vectorizer.transform(X_train)
X_test_vec = vectorizer.transform(X_test)
X_train_vec.get_shape()

(3927, 9132)

We can see that each of our sentences is now represented by a vector of length 9132. It might be interesting to compare text and numeric representation:

In [637]:
# Compare original comment text with its numeric vector representation
print(f"Original sentence:\n{X_train[3:4].values}\n")
# Feature Matrix
features = pd.DataFrame(
    X_train_vec[3:4].toarray(), columns=vectorizer.get_feature_names()
)
nonempty_feat = features.loc[:, (features != 0).any(axis=0)]
print(f"Vector representation of sentence:\n {nonempty_feat}")

Original sentence:
['Absolut zufrieden mit dem Team , von dem ich am 30.09.2011 um 13.00 ambulant operiert wurde. Die Krankenschwester waren auch sehr gut. Danke!']

Vector representation of sentence:
          00        09        13      2011        30  absolut        am  \
0  0.185898  0.214235  0.196529  0.201964  0.161042  0.15359  0.092103   

   ambulant  auch sehr    danke       dem   dem ich  dem team  \
0  0.198253   0.171252  0.13431  0.173229  0.204667  0.237733   

   die krankenschwester    ich am  krankenschwester   mit dem  operiert  \
0              0.227242  0.197669          0.185069  0.131094  0.129453   

   operiert wurde  sehr gut      team  team von        um     um 13   von dem  \
0        0.245439  0.109134  0.127542  0.227242  0.098073  0.258141  0.194352   

      waren  waren auch  wurde die  zufrieden  zufrieden mit  
0  0.092489     0.22125   0.192297   0.130723       0.206097  


For this five word sentence, the vector of length  9132 contains mostly zeros. However, the indices representing the used words / ngrams are non empty. They include the value that TF-IDF assigned to them. In this particular case, "um 13" (at 13 o'Clock) has the largest weight meaning that it is relatively frequent in our sentence while not being very common in other sentences of our dataset.

Now, that we have prepared our features, we can start to train and evaluate models. For a binary classification task there are many options to chose from in scikit-learn. We will focus on the ones that are most promising. In my experience they are: Logistic Regression, Support Vector Classification (SVC), Ensemble Methods (Boosting, Random Forest) and Neural Networks (i.e. Multi Layer Perceptron or MLP in sklearn). We will compare these models and chose the most promising one:

In [640]:
# models to test
classifiers = [
    LogisticRegression(solver="sag", random_state=1),
    LinearSVC(random_state=1),
    RandomForestClassifier(random_state=1),
    XGBClassifier(random_state=1, use_label_encoder =False),
    MLPClassifier(
        random_state=1,
        solver="adam",
        hidden_layer_sizes=(12, 12, 12),
        activation="relu",
        early_stopping=True,
        n_iter_no_change=1,
    ),
]
# get names of the objects in list (too lazy for c&p...)
names = [re.match(r"[^\(]+", name.__str__())[0] for name in classifiers]
print(f"Classifiers to test: {names}")

Classifiers to test: ['LogisticRegression', 'LinearSVC', 'RandomForestClassifier', 'XGBClassifier', 'MLPClassifier']


In [641]:
%%time
# test all classifiers and save pred. results on test data
results = {}
for name, clf in zip(names, classifiers):
    print(f"Training classifier: {name}")
    clf.fit(X_train_vec, Y_train)
    prediction = clf.predict(X_test_vec)
    report = sklearn.metrics.classification_report(Y_test, prediction)
    results[name] = report

Training classifier: LogisticRegression
Training classifier: LinearSVC
Training classifier: RandomForestClassifier
Training classifier: XGBClassifier
[12:40:16] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
Training classifier: MLPClassifier
Wall time: 17.5 s


Training LogisticRegression and LinearSVC was very fast while the remaining classifiers were significantly slower. This has to do with their higher model complexity but can also greatly vary depending on the parameters used.
After having trained all our models on the train data and applying their prediction on the test data, we can judge their performance:

In [642]:
# Prediction results
for k, v in results.items():
    print(f"Results for {k}:")
    print(f"{v}\n")

Results for LogisticRegression:
              precision    recall  f1-score   support

           0       0.94      0.89      0.91       515
           1       0.93      0.96      0.95       794

    accuracy                           0.93      1309
   macro avg       0.93      0.93      0.93      1309
weighted avg       0.93      0.93      0.93      1309


Results for LinearSVC:
              precision    recall  f1-score   support

           0       0.91      0.92      0.91       515
           1       0.95      0.94      0.94       794

    accuracy                           0.93      1309
   macro avg       0.93      0.93      0.93      1309
weighted avg       0.93      0.93      0.93      1309


Results for RandomForestClassifier:
              precision    recall  f1-score   support

           0       0.90      0.88      0.89       515
           1       0.92      0.93      0.93       794

    accuracy                           0.91      1309
   macro avg       0.91      0.91  

The output of sci-kit's classification_report provides us with several metrics. For unbalanced data sets accuracy is an inappropriate metric. Since it's value solely tells you how many cases have been properly classified, a high value can be achieved by always predicting the majority class. In contrast, the f1-score puts precision and recall of each class in relation to each other. As such, it is a more fine grained measure. We can use an aggregate version of it to have a single metric summarizing the performance of each model. For example, the weighted f1-score is an average of the f1-scores of both our classes taking the class distribution into account. On the other hand, the macro f1-score averages over class scores without weighing them. Consequently, we'll use that as we wish to give the same importance to both our classes. This is because even though bad grades are much more rare they also have a more severe impact.
All methods achieve impressive results predicting the good ratings class with f1-scores above 0.89. However, results for the bad ratings class are much lower and vary wildly.
Here, the ensemble methods, i.e. RandomForest and XGBoost, perform worst. However, with some more effort to tune their parameters they would probably fare significantly better.
As a general rule of thumb, logistic regressions deliver decent results in many different use cases. Moreover, they are simple to apply, robust and computationally efficient. macro f1-score we have three best score, MLP, LR and SVC over 0.92. In general, SVC is comparable in speed and simplicity to the logistic regression. In addition, it often performs very well in classification tasks as we can see here. In contrast, Neural Networks perform extremely well in all sorts of tasks but are also much more complex and slow all around. Because of this, we will stick to the LinearSVC classifier for now.

### Parameter tuning

We've learned that linear SVC is a solid model choice and delivers great results out of the box. Still, we might be able to improve upon those by taking a more guided approach to choosing parameters. To do so, we will compare different parameters for feature creation as well as modeling. We can achieve this by making use of the pipeline and grid search functionality in sci-kit learn.  
The `Pipeline` object encapsulates several processing steps into one. The last step in a pipeline must have a `fit()` functionality. The previous steps a `transform()` and `fit()` functionality. The `Pipeline` object itself has the same `fit()`, `transform()` and `predict()` methods as any other model in scikit-learn. Thus, we can streamline the whole process of feature creation, model fitting and prediction into one step.  
With `GridSearchCV` we can define parameter spaces for our functions. All parameter combinations will be evaluated against one another in a cross validation using a defined score metric. Next, we combine our pipeline with grid search. In result, we can jointly test combinations of parameters in feature creation and model training:

In [643]:
# feature creation and modelling in a single function
pipe = Pipeline([("tfidf", TfidfVectorizer()), ("svc", LinearSVC())])

# define parameter space to test # runtime 35min
params = {
    "tfidf__ngram_range": [(1, 1), (1, 2), (1, 3)],
    "tfidf__max_df": np.arange(0.3, 0.8, 0.2),
    "tfidf__min_df": np.arange(5, 100, 45),
}
pipe_clf = GridSearchCV(pipe, params, n_jobs=-1, scoring="f1_macro")
pipe_clf.fit(X_train, Y_train)
pickle.dump(pipe_clf, open("./clf_pipe.pck", "wb"))

Each value added to the `params` to be tested with `GridSearchCV` increases the run time of the training significantly. To limit the combinations, we first optimize the feature creation step. We test for different values for `ngram`. Moreover, `max_df` and `min_df` set an upper and lower limit for word frequencies. We want to exclude infrequent words because the model won't be able to learn (meaningful) associations with very few observations. The same is true for very frequent words which won't allow the model to differentiate between classes. For each parameter we check the values which returned the best model fit:

In [644]:
print(pipe_clf.best_params_)

{'tfidf__max_df': 0.3, 'tfidf__min_df': 5, 'tfidf__ngram_range': (1, 3)}


Now, using this best parameters for TF-IDF we can search for optimal parameters for the LinearSVC classifier:

In [645]:
%%time
# feature creation and modelling in a single function
pipe = Pipeline([("tfidf", TfidfVectorizer()), ("svc", LinearSVC())])

# define parameter space to test # runtime 19min
params = {
    "tfidf__ngram_range": [(1, 3)],
    "tfidf__max_df": [0.5],
    "tfidf__min_df": [5],
    "svc__C": np.arange(0.2, 1, 0.15),
}
pipe_svc_clf = GridSearchCV(pipe, params, n_jobs=-1, scoring="f1_macro")
pipe_svc_clf.fit(X_train, Y_train)
pickle.dump(pipe_svc_clf, open("./pipe_svc_clf.pck", "wb"))

Wall time: 44 s


In [646]:
best_params = pipe_svc_clf.best_params_
print(best_params)

{'svc__C': 0.6499999999999999, 'tfidf__max_df': 0.5, 'tfidf__min_df': 5, 'tfidf__ngram_range': (1, 3)}


We focus on the C parameter which is basically a regularization parameter and essential in the performance of the SVC classifier. High values of C mean the margin of the hyperplane chosen by SVC to separate the data will be smaller. Thus, while classification on training data will be better this can also lead to overfitting. Consequently, C controls a trade off between a low training and testing error.

Finally, we combine these best parameters and test the prediction of our model using the pipe:

In [647]:
# run pipe with optimized parameters
pipe.set_params(**best_params).fit(X_train, Y_train)
pipe_pred = pipe.predict(X_test)
report = sklearn.metrics.classification_report(Y_test, pipe_pred)
print(report)

              precision    recall  f1-score   support

           0       0.93      0.92      0.93       515
           1       0.95      0.95      0.95       794

    accuracy                           0.94      1309
   macro avg       0.94      0.94      0.94      1309
weighted avg       0.94      0.94      0.94      1309



Using parameter tuning we have slightly improved on our already decent model by increasing the precision for class 1.0. However, we can see that the margin for improvement is little. On one hand, this is because our initial parameters were already very close to the optimum. On the other hand, it might be that given our data and the approach used we might already be close to a barrier.

### Prediction

Now, that we have our best performing model (so far) we can use it to make predictions. One possible application is to find contradictory reviews, i.e. reviews where the sentiment of the comment doesn't match the rating. For that, we look at cases where our model makes a prediction with high confidence which doesn't match the original rating:

In [648]:
# Get confidence score for prediction
conf_score = pipe.decision_function(X_test)
# Get the Nth highest / lowest score
# high score indicates class 1 (bad), low score 0 (good)
score_neg = np.sort(conf_score)[-1300]
score_pos = np.sort(conf_score)[1300]
print(
    f"Threshold for negative rating: {score_neg}\nThreshold for positive rating: {score_pos}"
)

Threshold for negative rating: -1.512803611456657
Threshold for positive rating: 1.9890163295026648


The `decision_function` returns values ranging from negative to positive (without a general boundary). They depict the distance of the input to the hyperplane which the SVM algorithm uses to separate the two classes. In our case, large values indicate a higher confidence in class 1 (i.e. a bad rating) and low values in class 0. Consequently, we take the n highest (lowest) values as a threshold for a high confidence in class 1 (0). 

In [649]:
pd.options.display.max_colwidth = 800
# Predicted good but rated bad
test[["Rating", "Textuelle Bewertung"]][(Y_test != pipe_pred) & (conf_score <= score_pos)]

Rating  \
314        0   
5130       1   
5194       0   
3767       1   
4879       0   
4517       0   
4238       1   
1832       1   
4899       0   
5116       1   
1304       1   
1297       1   
5063       0   
1206       1   
2234       1   
2018       0   
4455       1   
795        0   
1159       0   
1413       0   
2253       1   
1885       1   
4839       0   
2706       0   
1495       0   
440        0   
2807       1   
1198       0   
1160       1   
331        0   
4998       0   
4120       1   
1414       0   
354        0   
12         1   
4642       0   
4515       1   
244        0   
517        0   
4995       0   
4835       0   
2479       1   
4326       0   
2249       0   
324        0   
301        0   
2274       1   
4176       0   
2567       0   
4216       1   
2327       0   
4598       0   
808        1   
2118       1   
4349       0   
1675       1   
4200       1   
827        1   
190        1   
593        1   
1078       0   
40         1   
306        1   
5146       1   
4241       0   
2471       1   
4773       1   
2555       0   
1497       1   
4608       0   
4872       0   
700        1   
770        0   
1235       1   
2517       0   
5144       1   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  Textuelle Bewertung  
314                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
5130                                                                                                                                                                                                                                                                                                                                                                                                       Es kommt immer drauf an, wann man kommt und was man hat bzw. zu welchem Arzt man kommt. Ich habe in der Notaufnahme schon tolle, kompetente Ärzte gehabt. Aber auch schon Ärzte, die kaum ein Wort deutsch sprachen, das Problem nicht verstehen und keine Lösung gefunden haben oder eine später erkennbar falsche Diagnose gestellt haben. An einer durchgehenden Qualität sollte meiner Meinung nach gearbeitet werden.  
5194                                                                                                                                                                                                                                                                                                                                                         

In [650]:
# Predicted bad but rated good
test[["Rating", "Textuelle Bewertung"]][(Y_test != pipe_pred) & ((conf_score >= score_neg))]

Rating  \
314        0   
5130       1   
5194       0   
3767       1   
4879       0   
4517       0   
4238       1   
1832       1   
4899       0   
5116       1   
1304       1   
1297       1   
5063       0   
1206       1   
2234       1   
2018       0   
4455       1   
795        0   
1159       0   
1413       0   
2253       1   
1885       1   
4839       0   
2706       0   
1495       0   
440        0   
2807       1   
1198       0   
1160       1   
331        0   
4998       0   
4120       1   
1414       0   
354        0   
12         1   
4642       0   
4515       1   
244        0   
517        0   
4995       0   
4835       0   
2479       1   
4326       0   
2249       0   
324        0   
301        0   
2274       1   
4176       0   
2567       0   
4216       1   
2327       0   
4598       0   
808        1   
2118       1   
4349       0   
1675       1   
4200       1   
827        1   
190        1   
593        1   
1078       0   
40         1   
306        1   
5146       1   
4241       0   
2471       1   
4773       1   
2555       0   
1497       1   
4608       0   
4872       0   
700        1   
770        0   
1235       1   
2517       0   
5144       1   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  Textuelle Bewertung  
314                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
5130                                                                                                                                                                                                                                                                                                                                                                                                       Es kommt immer drauf an, wann man kommt und was man hat bzw. zu welchem Arzt man kommt. Ich habe in der Notaufnahme schon tolle, kompetente Ärzte gehabt. Aber auch schon Ärzte, die kaum ein Wort deutsch sprachen, das Problem nicht verstehen und keine Lösung gefunden haben oder eine später erkennbar falsche Diagnose gestellt haben. An einer durchgehenden Qualität sollte meiner Meinung nach gearbeitet werden.  
5194                                                                                                                                                                                                                                                                                                                                                         

In [693]:
# Get new comments from website that were not included in original data
INPUT = ["Nach diesen Erfahrungen fühlten wir uns sehr hilflos und allein gelassen und fragen uns wie ernst der Sozialdienst dieses Klinikums seine Arbeit nimmt .fazit: Nie wieder dieses Klinikum !!"]

   
"""Sprechstundenhilfe war super nett man fühlt sich wohl.",
     "Super sympathische Ärztin, fühle mich bei ihr bestens aufgehoben."
    "Frau Doktor Merz nimmt sich richtig Zeit für mich. Hilft wo sie kann."
    "Hört wirklich einen zu. Sehr nett und freundlich. Sie ist sehr kompetent,"
    "zuverlässig und vertrauenswürdig.",
    "Nach meiner Beobachtung hat diese Praxis eine schlechte Hygiene. ",
    "Mangels akriebischer Behandlung musste mehrmals nachgebessert werden."""

'Sprechstundenhilfe war super nett man fühlt sich wohl.",\n     "Super sympathische Ärztin, fühle mich bei ihr bestens aufgehoben."\n    "Frau Doktor Merz nimmt sich richtig Zeit für mich. Hilft wo sie kann."\n    "Hört wirklich einen zu. Sehr nett und freundlich. Sie ist sehr kompetent,"\n    "zuverlässig und vertrauenswürdig.",\n    "Nach meiner Beobachtung hat diese Praxis eine schlechte Hygiene. ",\n    "Mangels akriebischer Behandlung musste mehrmals nachgebessert werden.'

In [694]:
# Pre-Process comments as we did with train data
text = [clean_text(comment) for comment in INPUT]
text_out = " \n".join(text)
print(f"Input after pre-processing / cleaning:\n\n{text_out}")
# run comments through pipe: predict using our best model from above
predictions = pipe.predict(text)

Input after pre-processing / cleaning:

erfahr fuhlt hilflos allein gelass frag ernst sozialdien klinikum arbeit nimmt fazit nie klinikum


In [695]:
# Show comment and predicted Labels
predictions = pd.Series(predictions)
predictions = predictions.replace(0, "Schlechte Erfahrung").replace(1, "Gute Erfahrung")

pd.concat(
    [pd.Series(INPUT), predictions], axis="columns", keys=["Kommentar", "Hervorsage"]
)

Kommentar  \
0  Nach diesen Erfahrungen fühlten wir uns sehr hilflos und allein gelassen und fragen uns wie ernst der Sozialdienst dieses Klinikums seine Arbeit nimmt .fazit: Nie wieder dieses Klinikum !!   

            Hervorsage  
0  Schlechte Erfahrung